metaflow/plugins/kubernetes/kubernetes_runner.py

```diff
@@ -0,0 +1,216 @@
+import json
+import random
+import string
+import sys
+from datetime import datetime, timedelta
+
+from metaflow.datastore.util.s3tail import S3Tail
+from metaflow.mflog import delayed_update_while
+from metaflow.mflog.mflog import refine
+from metaflow.plugins.aws.utils import TaskSpec, build_task_command
+
+if sys.version_info > (3, 0):
+    from typing import TYPE_CHECKING, Callable, Dict, List, Optional
+    from datetime import timezone
+
+    if TYPE_CHECKING:
+        from metaflow.metaflow_environment import MetaflowEnvironment
+
+from metaflow.metaflow_config import (
+    DATASTORE_SYSROOT_S3,
+    DATATOOLS_S3ROOT,
+    DEFAULT_METADATA,
+    METADATA_SERVICE_HEADERS,
+    METADATA_SERVICE_URL,
+)
+
+
+def generate_job_name(step_name):
+    return (
+        "".join([random.choice(string.ascii_lowercase) for _ in range(5)])
+        + "-"
+        + step_name
+    )
+
+
+def create_job_object(
+    job_name,  # type: str
+    image,  # type: str
+    command,  # type: List[str]
+    env,  # type: Dict[str, str]
+    labels,  # type: Dict[str, str]
+):
+    from kubernetes import client
+
+    # Configureate Pod template container
+    container = client.V1Container(
+        name=job_name,
+        image=image,
+        command=command,
+        env=[client.V1EnvVar(name=k, value=v) for k, v in env.items()],
+    )
+    # Create and configurate a spec section
+    template = client.V1PodTemplateSpec(
+        metadata=client.V1ObjectMeta(labels={"app": "pi"}),
+        spec=client.V1PodSpec(restart_policy="Never", containers=[container]),
+    )
+    # Create the specification of deployment
+    spec = client.V1JobSpec(
+        template=template,
+        backoff_limit=1,  # retries handled by Metaflow
+        ttl_seconds_after_finished=600,
+    )
+
+    # Instantiate the job object
+    job = client.V1Job(
+        api_version="batch/v1",
+        kind="Job",
+        metadata=client.V1ObjectMeta(name=job_name, labels=labels),
+        spec=spec,
+    )
```

Comment:
k8s python client accepts specifications written as dictionaries, then it looks identical to the k8s json spec.  The job could be created like this:
```
job = {
    "apiVersion": "batch/v1",
    "kind": "Job",
    "metadata": {
        "name": job_name,
        "labels": labels
    },
    "spec": {
        "template": {
            "metadata": {
                "labels": {
                    "app": "pi",
                },
            },
            "spec": {
                "containers": [
                    {
                        "name": job_name,
                        "image": image,
                        "command": command,
                        "env": [{"name": k, "value": v} for k, v in env.items()],
                    },
                ],
                "restartPolicy": "Never"
            }
        },
        "backoffLimit": 1,      # retries handled by Metaflow
        "ttlSecondsAfterFinished": 600
    }
}
```

Reply:
Nice, didn't know that. One thing I like about `client.` API is that VSCode/mypy complains when I make a typo in a parameter name. For example I always forget that `"template" {...}` bit when writing YAML/JSON.